Multisite template (Ema 0.8 upgrade)

srid opened the pull request:
Porting for https://github.com/srid/ema/pull/81 

Resolves #20 

- [x] Tailwind CSS cache issue in static site (live-server: no issue, due to use of UUID and CSS changing only on .hs change)

srid committed:
Use multisite refactor of Ema, and new unionmount

Split sites into two.

srid committed:
Merge remote-tracking branch 'origin/master' into multisite

srid committed:
Update ema, for new API

srid committed:
Update ema -- mega update

srid committed:
Update ema

srid committed:
Muiti site demo

srid committed:
cleanup

srid committed:
Add hercules CI setting to flake.nix

srid committed:
Update ema

srid committed:
direnv

srid committed:
Add treefmt

srid committed:
enable fourmolu

srid committed:
cleanup flake

srid committed:
Add fourmolu config

srid committed:
rm tmux

srid committed:
note about static site url hack

srid committed:
Redundant

srid committed:
Update ema, etc.

srid committed:
Sync with haskell-template

srid committed:
format cabal file

srid committed:
nix: remove path-tree (now on hackage)

srid committed:
unionmount on Hackage

srid committed:
remove hercules

srid committed:
ci: cleanup

srid committed:
static: SOP not needed

srid committed:
Pass SiteArg to both apps, removing hardcoding

srid committed:
nix: Update all inputs

srid committed:
cabal: Add -threaded (resolves #23)

srid committed:
Update ema, and use haskell-flake for nix

srid committed:
Redundant

srid committed:
Some doc touches

srid committed:
Clean up a bit, and use unionmount upstream

srid committed:
Revamp static file handling

srid committed:
Remove metadata stuff

srid committed:
fix static links

srid committed:
remove format

srid committed:
fix index.md

srid committed:
no flake-compat

srid committed:
Add tailwind.css to repo

srid committed:
mark as autogenerated

srid committed:
Precise route encoder; handle missing .md routes

srid committed:
cleanups

srid committed:
remove breadcrumbs; other cleanups

srid committed:
Use combineRouteEncoder

srid commented on src/Main.hs:
```diff
@@ -80,195 +88,185 @@ markdownRouteFileBase :: MarkdownRoute -> Text
+  allRoutes =
+    toList
+
+-- Route encoders can also be composed. Here, we use `combineRouteEncoder` to
+-- combine two route encoders.
+instance IsRoute Route where
+  type RouteModel Route = Model
+  routeEncoder =
+    combineRouteEncoder
+      routeDecomposition
+      decomposeModel
+      (routeEncoder @MarkdownRoute)
+      (routeEncoder @StaticRoute)
+  allRoutes model =
+    fmap Route_Markdown (allRoutes @MarkdownRoute $ fst . decomposeModel $ model)
+      <> fmap Route_Static (allRoutes @StaticRoute $ snd . decomposeModel $ model)
+
+decomposeModel :: Model -> (Set MarkdownRoute, Set FilePath)
+decomposeModel = Map.keysSet . modelDocs &&& modelFiles
+
+routeDecomposition :: Iso' (Either MarkdownRoute StaticRoute) Route
+routeDecomposition =
+  iso (either Route_Markdown Route_Static) $ \case
+    Route_Markdown r -> Left r
+    Route_Static r -> Right r
```
Too annoying to have the user do this. Ema should be able to generically derive this. https://github.com/srid/ema/pull/81#issuecomment-1163226891

srid replied:
resolved in [f76280e](https://github.com/srid/ema-template/pull/22/commits/f76280e5189457a1ead3f2855bd03e62ca8cb8fc)

srid committed:
Advance Ema, to use new generics

srid committed:
Introduce SlugListRoute (decomposes better)

srid committed:
More safer static URL handling

srid commented on src/Main.hs:
```diff
@@ -37,16 +37,41 @@ import Text.Pandoc.Builder qualified as B
+        )
+  deriving (IsRoute) via (Route `WithModel` Model)
+
+instance HasSubModels Route where
+  subModels m =
+    I (Set.map coerce $ Map.keysSet $ modelDocs m) :* I (modelFiles m) :* Nil
```
By the way, @lucasvreis - note here that the first sub-model (for markdown routes) is not really a sub-type (ie., it is not merely a field of the larger model) ... but rather an *arbitrary computation* (`Set.map coerce . Map.keysSet`) derived from that field.

We can conclude then that not all sub-models are sub-types.

lua-vr replied:
Well, you could define `SlugMapRoute` on top of `SlugListRoute`, and this arbitary computation would then belong in `SlugMapRoute`

srid committed:
Remove uuid hack; store lastAccessed in StaticRoute model

The lastAccessed time is used to force-reload CSS and such in the
browser. Keeping it in model is nifty.

srid committed:
Update ema

srid committed:
vscode: add manageHLS

srid committed:
Add hlint.yaml

srid committed:
Add bin/hoogle

srid committed:
Apply hlints

srid committed:
Update ema, and use StaticRoute

srid committed:
Update ema

srid committed:
Rewrite template using basic example

srid committed:
Merge pull request #24 from EmaApps/back-to-basics

Rewrite template using basic example

srid committed:
Use GHC 9.2.3 (#25)

* Add/use garnix cache

* nix: Use GHC 9.2.3

* Drop WithSubRoutes, because ghc 9.2

* Update ema, and use TH

cf. https://github.com/EmaApps/ema/issues/103

* streaming-commons

* http2 too

* readme: mention garnix

* garnix: disable M1 devShell

srid committed:
Advance ema

srid committed:
Simplify shell.nix

srid committed:
nix: Update inputs

srid committed:
Update ema, and nixpkgs

srid committed:
mv DAC

srid committed:
Re-enable cabal-fmt

srid committed:
Add cli argument handling, including setting of baseurl

srid committed:
use baseurl in ci

srid committed:
nix run tailwind

srid committed:
simplify

srid committed:
use ema master branch
